packages/core/src/contrib/Constraints.ts

```diff
@@ -468,18 +569,68 @@ export const containsGroupShape = (
   clip: ClipData<ad.Num>,
   s2: Shape<ad.Num>,
   padding: ad.Num
-): ad.Num => {
-  const vals = shapes.map((s) => constrDict.contains.body(s, s2, padding));
+): MayWarn<ad.Num> => {
+  const results = shapes.map((s) => contains(s, s2, padding));
+  const energies = results.map((r) => r.value);
+  const warningss = results.map((r) => {
+    return r.warnings.map((warning) => {
+      if (warning.tag === "BBoxApproximationWarning") {
+        const newStack: [
+          BBoxApproximationWarningItem,
+          ...BBoxApproximationWarningItem[]
+        ] = [
+          ...warning.stack,
+          {
+            signature: `containsGroupShape([${shapes
+              .map((s) => s.shapeType)
+              .join(", ")}], ..., ${s2.shapeType}, ...)`,
+          },
+        ];
+        return {
+          ...warning,
+          stack: newStack,
+        };
+      } else {
+        return warning;
+      }
+    });
+  });
   if (clip.tag === "NoClip") {
     // If a group does not have a clipping shape, checking whether a group contains a shape is equivalent to checking whether some group member contains the shape.
-    return minN(vals);
+    return {
+      value: minN(energies),
+      warnings: warningss.flat(),
+    };
   } else {
     // Otherwise, then (1) the group members (excluding the clipping shape) contains the other shape, and
     // (2) the clipping shape contains the other shape.
-    return andConstraint(
-      minN(vals),
-      constrDict.contains.body(clip.contents, s2)
-    );
+    const clipResult = contains(clip.contents, s2);
+    const clipEnergy = clipResult.value;
+    const clipWarnings = clipResult.warnings.map((warning) => {
+      if (warning.tag === "BBoxApproximationWarning") {
+        const newStack: [
```

Comment:
The type is just `BBoxApproximationWarningItem[]` here?

Reply:
I'd like to make this more general. The idea is, a function warning might eventually have multiple variants even though we currently only have one. I'm putting this check here for this purpose.

Reply:
No, the type is `BBoxApproximationWarning`. This type is a variant of `FunctionWarning` which is a variant of `StyleWarning`. `BBoxApproximationWarningItem` is a single entry in the stack of `BBoxApproximationWarning`.